agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "php_call.h"
+#include "php_user_instrument.h"
+#include "php_datastore.h"
+#include "php_execute.h"
+#include "php_hash.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "nr_datastore_instance.h"
+#include "nr_segment_datastore.h"
+#include "util_logging.h"
+#include "util_memory.h"
+#include "util_strings.h"
+#include "util_sleep.h"
+
+/*
+ * Purpose : Return a copy of Monolog\Logger::API.
+ *
+ * Params  : 1. An instance of Monolog\Logger.
+ *
+ * Returns : The Monolog version number as an integer
+ */
+static int nr_monolog_version(zval* logger TSRMLS_DC) {
+  int retval = 0;
+  zval* api = NULL;
+  zend_class_entry* ce = NULL;
+
+  if (0 == nr_php_is_zval_valid_object(logger)) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger object is invalid", __func__);
+    return 0;
+  }
+
+  ce = Z_OBJCE_P(logger);
+  if (NULL == ce) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger has NULL class entry",
+                     __func__);
+    return 0;
+  }
+
+  api = nr_php_get_class_constant(ce, "API");
+  if (NULL == api) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: Logger does not have API", __func__);
+    return 0;
+  }
+
+  if (nr_php_is_zval_valid_integer(api)) {
+    retval = Z_LVAL_P(api);
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected API be a valid int, got type %d", __func__,
+                     Z_TYPE_P(api));
+  }
+
+  nr_php_zval_free(&api);
+  return retval;
+}
+
+/*
+ * Purpose : Convert $level argument of Monolog\Logger::addRecord to a string
+ * representation of Monolog's log level.
+ *
+ * Params  : Logger instance and Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log level name
+ */
+static char* nr_monolog_get_level_name(zval* logger,
+                                       NR_EXECUTE_PROTO TSRMLS_DC) {
+  zval* level_name = NULL;
+  char* level_name_string = nr_strdup("UNKNOWN");
+
+  if (!nr_php_object_has_method(logger, "getLevelName" TSRMLS_CC)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: Logger does not have getLevelName method", __func__);
+    return level_name_string;
+  }
+
+  zval* level = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  level_name = nr_php_call(logger, "getLevelName", level);
+  if (nr_php_is_zval_valid_string(level_name)) {
+    level_name_string = nr_strdup(Z_STRVAL_P(level_name));
```

Comment:
This seems like a potential memory leak since we [allocate mem via strdup](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/util_memory.c#L141)  in [line 75](https://github.com/newrelic/newrelic-php-agent/pull/489/files#diff-d48bf43b4158fad631ce65b0464841310be10cff57ac0b937061fb9d6c33bdc3R75), but then we allocate it again for this strdup.

Reply:
Nice catch - thank you! Good 👁️

Reply:
@zsistla This has been addressed in 39bd9db2ec30ad4422cb92a17ce09e73c84b62ca